### PR TITLE
Allow features to be disabled at build time

### PR DESCRIPTION
**Pull request by rjwats**
PR contains:

- **features.ini config file**: to allow features to be enabled/disabled as required
- **/rest/features endpoint**: to describe the available features to the UI
- **FeaturesContext and withFeatures HoC**: so the UI can dynamically disable features as required

Still WIP - main question is what else should be able to be disabled? Are OTA + NTP considered too "core" to be worth disabling?

### CODE REVIEW

File: features.ini [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[features]
+build_flags = 
+    -D FT_PROJECT=1
```
Comment (kasedy) [inline features.ini:3]: It is not clear what project means. I would rename it into FT_DEMO_PROJECT and moved /src dir into /lib/framework/demo_project

File: lib/framework/Features.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+// security feature on by default
+#ifndef FT_SECURITY
+#define FT_SECURITY 1
+#endif
+
+// mqtt feature on by default
+#ifndef FT_MQTT
+#define FT_MQTT 1
+#endif
```
Comment (kasedy) [inline lib/framework/Features.h:19]: We don't need to duplicate this macro variables. They are already defined in features.ini

File: lib/framework/StatefulService.h [hunk 3]
```diff
@@ -21,11 +21,11 @@ enum class StateUpdateResult {
   ERROR         // There was a problem updating the state, propagation should not take place
 };
 
-template <class T>
-using JsonStateUpdater = StateUpdateResult (*)(JsonObject& root, T& settings);
+template <typename T>
+using JsonStateUpdater = std::function<StateUpdateResult(JsonObject& root, T& settings)>;
```
Comment (rjwats) [inline lib/framework/StatefulService.h:25]: Looking for better ideas here, this increases memory usage but it's rather irritating not being able to use lambdas or std::bind.

File: lib/framework/Features.h (hunk 2 above)
Comment (rjwats) [reply lib/framework/Features.h:19]: Going to leave these in for now. I have ideas around wanting to be able to package a firmware up as an Arduino library (with WWWData.h included as a static resource) and this makes it easier to build without PlatformIO.

File: features.ini (hunk 1 above)
Comment (rjwats) [reply features.ini:3]: Will come back to this perhaps. The project is a "placeholder" rather than a "demo" and it's documented as "fair game" to change.

**Comment by rjwats** [comment]
> I don't need NTP in my project. Please consider to add disable flag for it. I believe that WiFi AP + WiFi STA + status are core pages. Everything else should be disabled

Good plan. I think I'll add OTA and NTP into the features tomorrow.

**Comment by rjwats** [comment]
> I don't need NTP in my project. Please consider to add disable flag for it. I believe that WiFi AP + WiFi STA + status are core pages. Everything else should be disabled

OTA and NTP features can be disabled now. README.md updated and I have removed many of the unnecessarry exclusion blocks.

It's probably good enough to merge, though I'd like to come back and look at the class hierarchy around SecuritySettingsService and SecurityManager to see if I can improve it - not a show stopper.

**Pull request merged**